service.js

```diff
@@ -30,8 +31,8 @@ class Service {
   }
 
   toJSON() {
-    const { id, type, serviceEndpoint } = this
-    return Object.assign({}, this, { id, type, serviceEndpoint })
+    const { id, type, serviceEndpoint, props } = this
+    return Object.assign({}, this, { id, type, serviceEndpoint, props })
```

Comment:
The object is being extended by `this`, which contains the already set "arbitrary" properties.

```js
return Object.assign({}, this, { id, type, serviceEndpoint })
```

The reason we do this is because the `id`, `type`, and `serviceEndpoint` properties are _getters_ on the _class prototype_  and therefore are not properties of the `this` object, but rather the `Service.prototype` object. You can test this with [`this.hasOwnProperty('serviceEndpoint')`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) and notice that it is `false`. [`Object.assign()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) will assign properties from the left to right to the `target` (first argument) object based on the getters and setters of the sources.